anymail/webhooks/postal.py

```diff
@@ -0,0 +1,169 @@
+    webhook_key = None
+
+    def __init__(self, **kwargs):
+        self.webhook_key = get_anymail_setting('webhook_key', esp_name=self.esp_name, kwargs=kwargs, allow_bare=True)
+
+        super(PostalBaseWebhookView, self).__init__(**kwargs)
+
+    def validate_request(self, request):
+        try:
+            signature = request.META["HTTP_X_POSTAL_SIGNATURE"]
+        except KeyError:
+            raise AnymailWebhookValidationFailure("X-Postal-Signature header missing from webhook")
+
+        public_key = serialization.load_pem_public_key(
+            ('-----BEGIN PUBLIC KEY-----\n' + self.webhook_key + '\n-----END PUBLIC KEY-----').encode(),
+            backend=default_backend()
+        )
+
+        try:
+            public_key.verify(
+                b64decode(signature),
+                request.body,
+                padding.PKCS1v15(),
+                hashes.SHA1()
+            )
+        except InvalidSignature:
+            raise AnymailWebhookValidationFailure(
+                "Postal webhook called with incorrect signature")
+
+
+class PostalTrackingWebhookView(PostalBaseWebhookView):
+    """Handler for Postal message, engagement, and generation event webhooks"""
+
+    signal = tracking
+
+    def parse_events(self, request):
+        esp_event = json.loads(request.body.decode("utf-8"))
+
+        raw_timestamp = esp_event.get("timestamp")
+        timestamp = (
+            datetime.fromtimestamp(int(raw_timestamp), tz=utc)
+            if raw_timestamp
+            else None
+        )
+
+        payload = esp_event.get("payload", {})
+
+        status_types = {
+            "Sent": EventType.DELIVERED,
+            "SoftFail": EventType.DEFERRED,
+            "HardFail": EventType.FAILED,
+            "Held": EventType.QUEUED,
+        }
+
+        def get_event_type(event):
+            if "status" in event:
+                return status_types.get(event["status"], EventType.UNKNOWN)
+
+            if "bounce" in event:
+                return EventType.BOUNCED
+
+            if "url" in event:
+                return EventType.CLICKED
+
+            return EventType.UNKNOWN
+
+        event_type = get_event_type(payload)
+
+        description = payload.get("details")
+        mta_response = payload.get("output")
+
+        # extract message-related fields
+        message = payload.get("message") or payload.get("original_message", {})
+        message_id = message.get("message_id")
+        tag = message.get("tag")
+        recipient = None
+        message_to = message.get("to")
+        if message_to is not None:
+            try:
+                recipient = parse_single_address(message_to).addr_spec
+            except AnymailInvalidAddress:
+                pass
+
+        if message.get("direction") == "incoming":
+            # this event is about an incoming email being delivered to Anymail
+            # we can safely ignore this
+            print("ignoring")
```

Comment:
You could raise an `AnymailConfigurationError` here, to help folks who have accidentally swapped Anymail webhook urls. ([Example](https://github.com/anymail/django-anymail/blob/v7.0/anymail/webhooks/postmark.py#L79-L82) from Postmark webhooks.)

Reply:
I noticed this in other implementations, but this case is a bit different. This is not the inbound email, but a tracking event _about_ the inbound email. So it's the right webhook, but it's not interesting to anymail users (and there's no `EventType` for it anyway).

Reply:
Oh, that's weird. This is the first ESP I've encountered that generates tracking events for inbound messages. (I'm not even really sure what it means to track a message you didn't send.) But yeah, probably not something we need to normalize for Anymail.